### PR TITLE
Create template_cloudinit_config resource

### PR DESCRIPTION
**Pull request by sthulb**
This is a branch from: https://github.com/hashicorp/terraform/pull/3416

@jen20 seems to have stopped development on his work for whatever reason, so I want to push it over the finishing line since it's of use to me and my company.

---

This does the exact same thing as @jen20's PR, I'm just finishing up his work.

- [x] Deal correctly with multi-part content header (not currently implemented)
- [x] Add resource tests for the config format
- [x] Fix up error messages to be reasonable and informative
- [x] Update documentation with resource
- [x] Document with GoDoc

### CODE REVIEW

**Comment by jen20** [comment]
:+1: I stopped working on this temporarily because I started working at HashiCorp and have been busy with core maintenance of Terraform instead of new features. I still think this is an important feature for Terraform though so I'm happy to see it continued and will review and merge as it matures.

File: CHANGELOG.md [hunk 1]
```diff
@@ -3,6 +3,7 @@
 FEATURES:
 
   * **New resource: `digitalocean_floating_ip`** [GH-3748]
+  * **New provider: `cloudinit` - Multipart cloudinit [GH-4061], [GH-3416]
```
Comment (jen20) [inline CHANGELOG.md]: There is no need to update the CHANGELOG on pull requests. We tend to find this results in merge conflicts, so upon merging we update it directly on the `master` branch.

File: builtin/providers/cloudinit/resource.go [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+func resource() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceCreate,
+		Delete: resourceDelete,
+		Exists: resourceExists,
+		Read:   resourceRead,
+
+		Schema: map[string]*schema.Schema{
+			"part": &schema.Schema{
+				Type:     schema.TypeSet,
```
Comment (jen20) [inline builtin/providers/cloudinit/resource.go]: My original implementation here was deficient - I think it turns out we can use `schema.TypeList` and then not require the ordering specification.

File: builtin/providers/cloudinit/resource.go [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+				Required: true,
+				ForceNew: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"order": &schema.Schema{
```
Comment (jen20) [inline builtin/providers/cloudinit/resource.go]: The need for this should be negated by using `schema.TypeList` instead of `schema.TypeSet`.

**Comment by jen20** [comment]
There are a few things we should consider here I think though after reviewing the code again.

- Rather than making this a new provider in it's own right, we should make it a sibling of `template_file` in the `template` provider.
- As I've commented in the diffs, we can likely use `schema.TypeList` instead of `schema.TypeSet` and clean up the need for order declaration substantially.

File: CHANGELOG.md (hunk 1 above)
Comment (sthulb) [inline CHANGELOG.md]: Fair enough, I'll remove it :)

File: CHANGELOG.md [hunk 4]
```diff
@@ -81,7 +81,7 @@ BUG FIXES:
   * core: Orphan resources can now be targets [GH-3912]
   * helper/schema: skip StateFunc when value is nil [GH-4002]
   * provider/google: Timeout when deleting large instance_group_manager [GH-3591]
-  * provider/aws: Fix issue with order of Termincation Policies in AutoScaling Groups. 
+  * provider/aws: Fix issue with order of Termincation Policies in AutoScaling Groups.
```
Comment (jen20) [inline CHANGELOG.md]: This CHANGELOG entry can also be reverted.

**Comment by jen20** [comment]
This is looking good!

**Comment by sthulb** [comment]
@jen20 Thanks, I'll take your comments under advisement.

Moving to template might make sense, as `template_multipart` rather than as `template_cloudinit`. I'll finish this up tomorrow hopefully.

**Comment by sthulb** [comment]
@jen20 I've completed the work for this PR now, other than the potential to merge the provider into `template`.

What do you think, leave as is or merge into `template`?

**Comment by jen20** [comment]
Hi @sthulb, this looks great. Glad to see it finally finished! I think we should merge it into the template provider, with the resource name `template_cloudinit_config` - I can finish that but if have time that would be great - I can likely get it merged today.

One last thing - would you mind rebasing your commits and squashing them? Just to keep a cleaner merge history. Feel free to force push on the branch.

**Comment by sthulb** [comment]
@jen20 All done :)

**Comment by jen20** [comment]
Great work! I'll review it now and get it merged in.

**Comment by jen20** [comment]
Hi @sthulb! Thanks for your work on this! I've cleaned up a couple of minor things and brought it into the main repository via #4095. However, since I did much of the original work on this, I'm going to get one of the other core team members to review it prior to merge. I'll close this PR in favour of #4095 so that it can be merged easily.

**Pull request closed without merging**